keps/sig-cli/2953-kustomize-plugin-graduation/README.md

```diff
@@ -0,0 +1,764 @@
+3. Run `kustomize edit generate-catalog`
+3. Run `kustomize build --trusted-plugin-catalog=catalog.yaml`
+
+```bash
+# Possible directory structure
+.
+⊢ Kustomization.yaml
+⊢ input.yaml
+⊢ reorder.yaml
+⊢ plugin-catalog.yaml
+∟ plugins
+  ∟ sorter
+    ⊢ sorter.rb
+    ∟ sorter_test.rb
+```
+
+```yaml
+# Example Kustomization.yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+resources:
+- input.yaml
+transformers:
+- reorder.yaml
+```
+
+```yaml
+# Example reorder.yaml
+apiVersion: local-config.example.co/v1
+kind: CustomSorter
+provider:
+  exec: 
+  	path: plugins/sorter/sorter.rb
+spec:
+  order: NsGKVN
+```
+
+```yaml
+# Example catalog.yaml (auto-generated)
+
+apiVersion: kustomize.io/v1
+kind: Catalog
+metadata:
+  name: "local-plugins"
+spec:
+  modules:
+  - apiVersion: local-config.example.co/v1
+    kind: CustomSorter
+    provider:
+      exec:
+        - path: plugins/sorter/sorter.rb
+          sha256: [a hash]
+```
+
+
+#### Story 2
+
+As an end user, I want to use one or more plugins published by a third party I trust.
+
+1. Write the config objects for the plugins you want to use, and reference them from your Kustomization's generators or transformers field as appropriate.
+2. Run `kustomize build --trusted-plugin-catalog=https://catalog.kpt.dev/1.2.3.json`
+
+
+```bash
+# Possible directory structure
+.
+⊢ Kustomization.yaml
+⊢ input.yaml
+⊢ conditionally-add-annotations.yaml
+∟ name-substring.yaml
+```
+
+```yaml
+# Example Kustomization.yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+catalogs:
+- https://catalog.kpt.dev/1.2.3.json
+resources:
+- input.yaml
+transformers:
+- name-substring.yaml
+- conditionally-add-annotations.yaml
+```
+
+```yaml
+# Example conditionally-add-annotations.yaml
+# Excerpted from https://catalog.kpt.dev/starlark/v0.2/
+apiVersion: fn.kpt.dev/v1alpha1
+kind: StarlarkRun
+metadata:
+  name: conditionally-add-annotations
+params:
+  toMatch:
+    config.kubernetes.io/local-config: "true"
+  toAdd:
+    configmanagement.gke.io/managed: disabled
+source: |
+  def conditionallySetAnnotation(resources, toMatch, toAdd):
+  ...
+```
+
+```yaml
+# Example name-substring.yaml
+# Excerpted from https://catalog.kpt.dev/ensure-name-substring/v0.1/
+apiVersion: v1
+kind: EnsureNameSubstring
+metadata:
+  name: my-fn-config
+substring: dev-
+```
+
+
+
+#### Story 3
+
+As a platform developer for a large organization, I want to provide internal users with a suite of trusted abstraction-based plugins.
+
+1. Develop the plugins in accordance with the [KRM Functions Specification] and publish them to one or more supported locations.
+2. Publish a Catalog (ideally using automation) to a supported location.
+3. Build a bespoke Kustomize distribution for internal use, e.g. `DEFAULT_TRUSTED_CATALOGS=kustomize.example.co/catalog/1.1.json make kustomize` (that hypothetical env var is build-time-only)
```

Comment:
curious if this is something that could be supported at runtime.  Some users will define ENV variables in their CICD pipelines and would be great to avoid retyping flags.

Reply:
Since we're ultimately talking about enabling third party code execution, I don't think it would be acceptable from a security perspective to make it settable via the environment.